Web Inspector: support awaitPromise parameter in Runtime.callFunctionOn

yury-s opened the pull request:
#### 700de83458a2ea7b582efa0fb8f50e748547847f
<pre>
Web Inspector: support awaitPromise parameter in Runtime.callFunctionOn
<a href="https://bugs.webkit.org/show_bug.cgi?id=264374">https://bugs.webkit.org/show_bug.cgi?id=264374</a>

Reviewed by Devin Rousso.

Added awaitPromise parameter to Runtime.callFunctionOn to allow getting
result of an asyncronouse function call with a single protocol command.
This allows to avoi possible navigation between Runtime.callFunctionOn command
and subsequent Runtime.awaitPromise which invalidates the function result
handle.

* Source/JavaScriptCore/inspector/InjectedScript.cpp:
(Inspector::InjectedScript::callFunctionOn):
* Source/JavaScriptCore/inspector/InjectedScript.h:
* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
(let.InjectedScript.prototype.callFunctionOn):
* Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.cpp:
(Inspector::InspectorRuntimeAgent::callFunctionOn):
* Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.h:
* Source/JavaScriptCore/inspector/protocol/Runtime.json:
* Source/WebCore/inspector/agents/page/PageRuntimeAgent.cpp:
(WebCore::PageRuntimeAgent::callFunctionOn):
* Source/WebCore/inspector/agents/page/PageRuntimeAgent.h:
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise-expected.txt: Added.
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html: Added.

Canonical link: <a href="https://commits.webkit.org/270784@main">https://commits.webkit.org/270784@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2de479d90e99bd30a04a5ab1213ef159f057ffe6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26378 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27640 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28496 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24145 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6791 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2419 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24130 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26638 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22685 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29076 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23659 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29714 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23000 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24058 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27609 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25594 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3458 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1677 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33039 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4881 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23407 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7140 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6338 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3936 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3782 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScript.cpp:
```diff
@@ -90,18 +90,16 @@ void InjectedScript::awaitPromise(const String& promiseObjectId, bool returnByVa
-    makeEvalCall(errorString, function, result, wasThrown, savedResultIndex);
-    ASSERT(!savedResultIndex);
+    function.appendArgument(awaitPromise);
+    makeAsyncCall(function, WTFMove(callback));
```
im a bit concerned with switching from `makeEvalCall` to `makeAsyncCall`

are we 100% sure that the behavior in the non-async (i.e. `awaitPromise: false`) case is the same?

yury-s replied:
Looking at the code, the non-waiting logic should be pretty much the same as before. The tests pass and manual testing doesn't show any regressions. I don't know how else we can validate that.

dcrousso replied:
i dont think a lot of the places we currently use `Runtime.callFunctionOn` are covered by tests :/

how much have you tried using Web Inspector with these changes?  AFAICT it seems like all of these user flows could be affected by this change:
- replaying 2D `<canvas>`
- "climbing" an object's prototype chain in the Console
- previewing `get` member functions of an object in the Console
- getting JavaScript completions in the Console
- displaying custom Audit result objects
- toggling CSS classes in the Styles details sidebar panel in the Elements Tab
- scrolling nodes into view via the Elements Tab
- etc.

yury-s replied:
OK, tried all these scenarios, didn't notice any regressions so far.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -207,14 +207,16 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
-        if (!isDefined(object))
-            return "Could not find object with given id";
+        if (!isDefined(object)) {
+            callback(this._createThrownValue("Could not find object with given id", objectGroupName));
```
i think this is actually a different behavior in that the protocol command would work (but return a synthetic error object) instead of erroring (with this string)

yury-s replied:
Yes, you are right. Updated this and the other place and added a couple tests.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -207,14 +207,16 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
-            return "Could not find object with given id";
+        if (!isDefined(object)) {
+            callback(this._createThrownValue("Could not find object with given id", objectGroupName));
+            return ;
```
style: unnecessary space

yury-s replied:
Removed.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
                     resolvedArgs[i] = this._resolveCallArgument(callArgs[i]);
                 } catch (e) {
-                    return @String(e);
+                    callback(this._createThrownValue(e, objectGroupName));
```
ditto for protocol command error behavior change

yury-s replied:
Done.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
-                "result", RemoteObject.create(func.@apply(object, resolvedArgs), objectGroupName, returnByValue, generatePreview),
-            );
+            if (typeof func !== "function") {
+                callback(this._createThrownValue("Given expression does not evaluate to a function", objectGroupName));
```
ditto for protocol command error behavior change

yury-s replied:
Done.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+                return;
+            }
+            let result = func.@apply(object, resolvedArgs);
+            if (awaitPromise && isDefined(result) && (InjectedScriptHost.internalConstructorName(result) === 'Promise')) {
```
`InjectedScriptHost.internalConstructorName` seems a bit overkill.  can't we just `result instanceof @Promise`?

Style: you can drop the `(` `)`
```
if (awaitPromise && isDefined(result) && result instanceof @Promise) {
```

yury-s replied:
`internalConstructorName` is necessary when the Promise is from another iframe. Added a test for that. Dropped `()` around ===.

dcrousso replied:
i coulda sworn that we create a new `InjectedScript` for each `JSGlobalObject`, so i dont think it'd be possible for this to be cross-realm in the first place.  do you have an example to the contrary?

if this is actually an issue then we probably should fix it for `awaitPromise` and `createFakeValueDescriptor` as well

Style: im still seeing the `(` `)`
Style: we prefer `"` instead of `'`

yury-s replied:
> i coulda sworn that we create a new InjectedScript for each JSGlobalObject, so i dont think it'd be possible for this to be cross-realm in the first place. do you have an example to the contrary?

`AwaitPromiseFromAnotherFrame` test demonstrates this. Basically each iframe has its own `Promise` instance so the types are technically different. You can try it in Web Inspector Console, open this URL in Safari:
```html
data:text/html,<iframe srcdoc='<script>window.iframePromise = Promise.resolve({b: 2023})</script>'></iframe>
```
and compare results of `frames[0].iframePromise instanceof Promise` and `frames[0].iframePromise instanceof frames[0].Promise` (should be false, true).


> if this is actually an issue then we probably should fix it for awaitPromise and createFakeValueDescriptor as well

Updated them too.

> Style: im still seeing the ( )

Oops, removed now.

> Style: we prefer " instead of '

Done.

dcrousso replied:
ah i see what you're getting at now.  even though we do create a unique `InjectedScript` for each `JSGlobalObject` you're trying to handle the case where access to the `<iframe>` is transparent because it's the same origin (or along those lines).

in that case what about if there's a custom `class Promise { }`?  perhaps we need to create some sort of `InjectedScriptHost.isPromise(value)` that checks `value->inherits<JSPromise>()`?

if we do need something like that, id personally suggest reverting these to `instanceof @Promise` for this PR and doing a followup PR to add `InjectedScriptHost.isPromise(value)`, but im also fine having it be part of this PR if you'd prefer

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+            }
+            let result = func.@apply(object, resolvedArgs);
+            if (awaitPromise && isDefined(result) && (InjectedScriptHost.internalConstructorName(result) === 'Promise')) {
+                result.then(value => {
```
Style: we always wrap arguments to arrow functions in `(` `)`
```
result.then((value) => {
```

yury-s replied:
Done.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+            let result = func.@apply(object, resolvedArgs);
+            if (awaitPromise && isDefined(result) && (InjectedScriptHost.internalConstructorName(result) === 'Promise')) {
+                result.then(value => {
+                    callback({
```
please use `@createObjectWithoutPrototype` to avoid issues with a modified/corrupted `Object.prototype`

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+                        wasThrown: false,
+                        result: RemoteObject.create(value, objectGroupName, returnByValue, generatePreview),
+                    });
+                }, reason => {
```
ditto
```
}, (reason) => {
```

yury-s replied:
Done.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+                    callback(this._createThrownValue(reason, objectGroupName));
+                });
+            } else {
+                callback({
```
ditto for using `@createObjectWithoutPrototype`

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+            }
         } catch (e) {
-            return this._createThrownValue(e, objectGroupName);
+            callback(this._createThrownValue(e, objectGroupName));
```
ditto for protocol command error behavior change

yury-s replied:
I believe here it was an error (non protocol) before the change too. Added tests that cover this.

dcrousso replied:
ah oops yes you're right

dcrousso commented on Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.cpp:
```diff
@@ -203,15 +200,15 @@ Protocol::ErrorStringOr<std::tuple<Ref<Protocol::Runtime::RemoteObject>, std::op
+            callback->sendSuccess(result.releaseNonNull(), WTFMove(wasThrown));
+    });
 
     if (pauseAndMute)
```
should this also move inside the lambda above?

or is that specifically not desired since other async things could happen in between whatever was `eval` and the `.then()`?

yury-s replied:
It is the latter. I don't think it is safe to mute console and skip exceptions for arbitrarily long time until the function is resolved.

dcrousso commented on Source/WebCore/inspector/agents/page/PageRuntimeAgent.cpp:
```diff
@@ -204,14 +204,16 @@ Protocol::ErrorStringOr<std::tuple<Ref<Protocol::Runtime::RemoteObject>, std::op
+        return;
+    }
 
     UserGestureEmulationScope userGestureScope(m_inspectedPage, emulateUserGesture.value_or(false), dynamicDowncast<Document>(executionContext(injectedScript.globalObject())));
```
ditto for whether this should be kept alive until the `.then()`

yury-s replied:
Here we just follow the same logic as if a human triggered the function: only the synchronous call would be within user gesture scope, while all actions triggered by it (e.g. `setTimeout` calls)  would lose that property.

dcrousso reviewed (changes requested):
although i have a few concerns, i overall think this is good :)

it definitely needs some tests tho

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,141 @@
+{
+    const expectedObject = {a: 2023};
+
+    const suite = InspectorTest.createAsyncSuite("Runtime.callFunction.awaitPromise");
```
Style: please use `let` for anything that isn't a true constant (e.g. `42`, `"test"`, `{a: 2023}`)

yury-s replied:
Done.

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,141 @@
+        description: "Check that callFunctionOn with awaitPromise=false works with a synchronous function.",
+        async test() {
+            const thisObject = await RuntimeAgent.evaluate.invoke({expression: `myObject`});
+            const result = await RuntimeAgent.callFunctionOn.invoke({functionDeclaration: `function() { return this; }`, awaitPromise: false, returnByValue: true, objectId: thisObject.result.objectId});
```
can we add some tests that have `awaitPromise: false, returnByValue: false` too?

Style: it would be a lot easier to read all of the long lines like this if we split it into separate lines
```
let result = await RuntimeAgent.callFunctionOn.invoke({
    functionDeclaration: `function() { return this; }`,
    awaitPromise: false,
    returnByValue: true,
    objectId: thisObject.result.objectId,
});
```

yury-s replied:
Done, added tests, reformatted the code.

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,141 @@
+        description: "Check that callFunctionOn with awaitPromise=false works with an asynchronous function.",
+        async test() {
+            const thisObject = await RuntimeAgent.evaluate.invoke({expression: `myObject`});
+            const result = await RuntimeAgent.callFunctionOn.invoke({functionDeclaration: `async function() { return Promise.reject(this); }`, awaitPromise: false, returnByValue: true, objectId: thisObject.result.objectId});
```
should this also `setTimeout(...)` like the previous test to really make sure it's not synchronous?

yury-s replied:
Done.

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,141 @@
+        async test() {
+            const thisObject = await RuntimeAgent.evaluate.invoke({expression: `myObject`});
+            const result = await RuntimeAgent.callFunctionOn.invoke({functionDeclaration: `async function() { return Promise.reject(this); }`, awaitPromise: false, returnByValue: true, objectId: thisObject.result.objectId});
+            InspectorTest.expectThat(!result.wasThrown, "Should not throw an error.");
```
NIT: id use `InspectorText.expectFalse` instead for this (and `InspectorText.expectTrue` for the other cases like this)

yury-s replied:
Updated.

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,141 @@
+            const objectId = JSON.stringify(parsedId);
+            try {
+                await RuntimeAgent.callFunctionOn.invoke({functionDeclaration: `async function() { await new Promise((fulfill, reject) => setTimeout(fulfill, 10)); return this; }`, awaitPromise: false, returnByValue: true, objectId});
+                InspectorTest.fail("Procol error was not thrown.");
```
oops typo `Procol`

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,141 @@
+            const thisObject = await RuntimeAgent.evaluate.invoke({expression: `myObject`});
+            try {
+                await RuntimeAgent.callFunctionOn.invoke({functionDeclaration: `1 + 2`, awaitPromise: false, returnByValue: true, objectId: thisObject.result.objectId});
+                InspectorTest.fail("Procol error was not thrown.");
```
ditto typo

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+                result.then((value) => {
+                    callback(@createObjectWithoutPrototype(
+                        "wasThrown", false,
+                        "result", RemoteObject.create(value, objectGroupName, returnByValue, generatePreview)
```
Style: missing trailing `,`

yury-s replied:
Done.

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
+            } else {
+                callback(@createObjectWithoutPrototype(
+                    "wasThrown", false,
+                    "result", RemoteObject.create(result, objectGroupName, returnByValue, generatePreview)
```
ditto

yury-s reviewed (commented):
Addressed all comments. PTAL.

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,263 @@
+                functionDeclaration: `function() { return this; }`,
+                awaitPromise: false,
+                returnByValue: true,
+                objectId: thisObject.result.objectId
```
Style: missing trailing comma

ditto below

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,263 @@
+                });
+                InspectorTest.fail("Protocol error was not thrown.");
+            } catch (e) {
+                InspectorTest.expectTrue(e.message.includes('Could not find object with given id'), "Received protocol error");
```
Style: we prefer `"` instead of `'`

ditto below

yury-s replied:
Dang, switching from one style to the another is not so easy. Updated all occurrences.

dcrousso commented on Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.cpp:
```diff
@@ -203,15 +200,15 @@ Protocol::ErrorStringOr<std::tuple<Ref<Protocol::Runtime::RemoteObject>, std::op
     }
 
-    injectedScript.callFunctionOn(errorString, objectId, functionDeclaration, arguments ? arguments->toJSONString() : nullString(), returnByValue.value_or(false), generatePreview.value_or(false), result, wasThrown);
+    injectedScript.callFunctionOn(objectId, functionDeclaration, arguments ? arguments->toJSONString() : nullString(), returnByValue.value_or(false), generatePreview.value_or(false), awaitPromise.value_or(false), [callback = WTFMove(callback)] (Protocol::ErrorString& errorString, RefPtr<Protocol::Runtime::RemoteObject>&& result, std::optional<bool>&& wasThrown, std::optional<int>&&) {
```
NIT: can you make the last parameter `std::optional<int>&& savedResultIndex` and then add a `UNUSED_PARAM(savedResultIndex)` so that it's more clear right here what it's for instead of having to look elsewhere to find out?

yury-s replied:
Done.

dcrousso commented on LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
```diff
@@ -0,0 +1,263 @@
+            let thisObject = await RuntimeAgent.evaluate.invoke({expression: `myObject`});
+            InspectorTest.deferOutputUntilTestPageIsReloaded();
+            let result = await RuntimeAgent.callFunctionOn.invoke({
+                functionDeclaration: `function() { window.location.reload(); return this; }`,
```
it seems like `window.location.reload` only actually schedules the reload rather than immediately doing it, so i guess that's why we're still able to get a JSON payload back from the inspected page

for what reason are we testing this?  generally speaking ive found that tests that reload the test page itself are often _very_ flaky, so im a bit hesitant about this

yury-s replied:
Yes, agreed about flakiness, removed the test. It would only make sense in comparison with `awaitPromise` as a separate command anyway.

dcrousso reviewed (approved):
awesome work!  a few things to fix/followup, but overall this looks great!  thanks for iterating with me 😁

yury-s committed:
Web Inspector: support awaitPromise parameter in Runtime.callFunctionOn
https://bugs.webkit.org/show_bug.cgi?id=264374

Reviewed by Devin Rousso.

Added awaitPromise parameter to Runtime.callFunctionOn to allow getting
result of an asyncronouse function call with a single protocol command.
This allows to avoi possible navigation between Runtime.callFunctionOn command
and subsequent Runtime.awaitPromise which invalidates the function result
handle.

* Source/JavaScriptCore/inspector/InjectedScript.cpp:
(Inspector::InjectedScript::callFunctionOn):
* Source/JavaScriptCore/inspector/InjectedScript.h:
* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
(let.InjectedScript.prototype.callFunctionOn):
* Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.cpp:
(Inspector::InspectorRuntimeAgent::callFunctionOn):
* Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.h:
* Source/JavaScriptCore/inspector/protocol/Runtime.json:
* Source/WebCore/inspector/agents/page/PageRuntimeAgent.cpp:
(WebCore::PageRuntimeAgent::callFunctionOn):
* Source/WebCore/inspector/agents/page/PageRuntimeAgent.h:
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise-expected.txt: Added.
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html: Added.

Canonical link: https://commits.webkit.org/270784@main
